### PR TITLE
Feature/implement join

### PR DESCRIPTION
**Pull request by romaninsh**
Implement $model->join()

 - [x] Implement basic join logic (class)
 - [x] Implement joins with "Array" persistence
 - [x] Implement joins for SQL

Reviewer:

 - [x] First review #26 
 - [ ] Check code
 - [ ] Don't mind if documentation is missing (will address later)

### CODE REVIEW

File: docs/joins.rst [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+you can even join using existing models::
+
+    $user->addField('username');
+    $user->addField('country_id');
+    $user->weakJoin('country')->importModel('Country');
+
+This will automatically import fields, expressions, relations and conditions from
+'Country' model into $user model and will also re-map field names in process. 
+
+.. php:method:: weakJoinModel
```
Comment (skondakov) [inline docs/joins.rst]: That's a new concept :)

File: phpunit.xml [hunk 2]
```diff
@@ -16,6 +16,8 @@
     <testsuites>
         <testsuite name="tests">
             <directory>tests</directory>
+            <exclude>tests/smbo</exclude>
```
Comment (skondakov) [inline phpunit.xml]: Should SMBO be here really?

File: tests/JoinArrayTest.php [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        $a = [];
+        $db = new Persistence_Array($a);
+        $m = new Model($db);
+
+        // tricky cases to testt
+        //
+        //$m->join('foo.bar', ['master_field'=>'baz']);
+        // foreign_table = 'foo.bar'
+    }
+}
```
Comment (skondakov) [inline tests/JoinArrayTest.php]: Really good test coverage

File: phpunit.xml (hunk 2 above)
Comment (romaninsh) [inline phpunit.xml]: yeah, I've added a few test-models, it won't have any data, but it should contain some interesting test-cases here. It should be OK as no sensitive data is shared.

File: tests/PersistentSQLTest.php [hunk 4]
```diff
@@ -51,4 +51,83 @@ public function testLoadArray()
...
+            'user'=>[
+                1=>['name'=>'John', 'surname'=>'Smith'],
+                2=>['name'=>'Sarah', 'surname'=>'Jones'],
+            ]
+        ];
+
+        $p = new Persistence_SQL('sqlite::memory:');
+
+        $p->connection->expr('drop table if exists user')->execute();
+        $p->connection->expr('create table user(id integer primary key autoincrement, name varchar(255), surname varchar(255))')->execute();
```
Comment (skondakov) [inline tests/PersistentSQLTest.php]: Is this a proper generic SQL syntax. AFAIK primary key cannot be set like this in all SQL implementations

File: tests/JoinArrayTest.php (hunk 3 above)
Comment (romaninsh) [inline tests/JoinArrayTest.php]: no worries, we'll work on test coverage once we have more functional lib.

File: tests/smbo/Model_Company.php [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            ['business_start', 'type'=>'date'],
+            ['director_name'],
+            ['vat_calculation_type','enum'=>['cash','invoice']],
+        ]);
+
+        $this->addFields([
+            ['is_vat_registered', 'type'=>'boolean'],
+        ]);
+    }
+}
```
Comment (skondakov) [inline tests/smbo/Model_Company.php]: Does SMBO tests belong here?

File: tests/PersistentSQLTest.php (hunk 4 above)
Comment (romaninsh) [inline tests/PersistentSQLTest.php]: this test is for sqlite which is invoked in line 102. Library supports other DB flavors but this particular tests uses sqlite, so should be OK.

**Comment by skondakov** [comment]
It has a good UT coverage. Please check the other comments and if solved and/or ok - it is ready for commit.

**Comment by romaninsh** [comment]
Yeey, nice!!

**Pull request merged**